addons/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Comm.java

```diff
@@ -0,0 +1,1145 @@
+ *
+ */
+
+public class KM200Comm<KM200BindingProvider> {
+
+    private static final Logger logger = LoggerFactory.getLogger(KM200Comm.class);
+    private HttpClient client = null;
+    private final KM200Device device;
+    public BigDecimal maxNbrRepeats;
+
+    public KM200Comm(KM200Device device) {
+        this.device = device;
+        maxNbrRepeats = BigDecimal.valueOf(10);
+        if (client == null) {
+            client = new HttpClient();
+        }
+    }
+
+    /**
+     * This function removes zero padding from a byte array.
+     *
+     */
+    private static byte[] removeZeroPadding(byte[] bytes) {
+        int i = bytes.length - 1;
+        while (i >= 0 && bytes[i] == 0) {
+            --i;
+        }
+        return Arrays.copyOf(bytes, i + 1);
+    }
+
+    /**
+     * This function adds zero padding to a byte array.
+     *
+     */
+    private static byte[] addZeroPadding(byte[] bdata, int bSize, String cSet) throws UnsupportedEncodingException {
+        int encrypt_padchar = bSize - (bdata.length % bSize);
+        byte[] padchars = new String(new char[encrypt_padchar]).getBytes(cSet);
+        byte[] padded_data = new byte[bdata.length + padchars.length];
+        System.arraycopy(bdata, 0, padded_data, 0, bdata.length);
+        System.arraycopy(padchars, 0, padded_data, bdata.length, padchars.length);
+        return padded_data;
+    }
+
+    /**
+     * This function does the GET http communication to the device
+     *
+     */
+    public byte[] getDataFromService(String service) {
+        byte[] responseBodyB64 = null;
+        int statusCode = 0;
+        // Create an instance of HttpClient.
+        synchronized (client) {
+            logger.debug("Starting receive connection...");
+
+            // Create a method instance.
+            GetMethod method = new GetMethod("http://" + device.getIP4Address() + service);
+
+            // Provide custom retry handler is necessary
+            method.getParams().setParameter(HttpMethodParams.RETRY_HANDLER,
+                    new DefaultHttpMethodRetryHandler(3, false));
+            // Set the right header
+            method.setRequestHeader("Accept", "application/json");
+            method.addRequestHeader("User-Agent", "TeleHeater/2.2.3");
+
+            try {
+                for (int i = 0; i < maxNbrRepeats.intValue() && statusCode != HttpStatus.SC_OK; i++) {
+                    // Execute the method.
+                    statusCode = client.executeMethod(method);
+                    // Check the status
+                    switch (statusCode) {
+                        case HttpStatus.SC_OK:
+                            break;
+                        case HttpStatus.SC_INTERNAL_SERVER_ERROR:
+                            /* Unknown problem with the device, wait and try again */
+                            logger.warn("HTTP GET failed: 500, internal server error, repeating.. ");
+                            Thread.sleep(100L * i + 1);
```

Comment:
Can you handle this (maybe on/from a higher level) without a `Thread.sleep` for example by (re)scheduling something in the future.

Reply:
This functions are only used inside a send or receive thread. So it is blocking only these and not the main thread. I would like to let it like it is.